028-service-binding.md

```diff
@@ -0,0 +1,374 @@
+
+The names for the keys given above match what is currently provided in other secrets created by Strimzi. The service binding specification does included some suggested fields for certificates, but they do not satisfy the requirements of an application connecting to Strimzi. These fields could be proposed back to the Service Binding specification.
+
+An alternative approach for naming is to name the fields to match the use, for example `ca` -> `truststore` (`truststore.crt`, `truststore.p12`, `truststore.password`), `user` -> `keystore` (`keystore.p12`, `keystore.password`, `keystore.crt`, `keystore.key`)
+ 
+
+### Implementation
+
+This proposal will result in a new `KafkaConnectionAssemblyOperator` class being added to the `cluster-operator` directory in the `strimzi-kafka-operator` repo. The class will be 
+deployed as a separate Verticle and have its own reconcile loop.
+
+The actions taken during the reconcile loop will be:
+1. On start-up the operator will watch all resources of Kind `KafkaConnection`.
+2. For any `KafkaConnection` resource present it will perform a GET request on the Kafka cluster that is 
+referenced in the `KafkaConnection` and pick out the configured listeners.
+3. If there are multiple listeners listed in the `Kafka` CR with the type `internal`, the operator 
+will filter the list by comparing the `tls` and `authentication` properties in the `Kafka` and `KafkaUser` CRs.
+4. If there are still multiple listeners it will pick one at random.
+5. The operator will add the name of the listener chosen will be added to the `KafkaConnection` status so that in future reconciles 
+the operator can take this into account in step 4 and avoid cycling through listeners.
+6. If there are no listeners marked as type `internal`, the reconcile loop will fail.
+7. The operator will then create a secret and add the bootstrapServers for the chosen listener to that secret.
+8. If the chosen listener uses TLS, the operator will then perform a GET request on the `cluster-ca-cert` secret and put the certificate and truststore files in the new secret.
+9. If the `KafkaConnection` references a `KafkaUser`, the operator will perform a GET request on that user to find the name of the secret containing the credentials and the username 
+from the status if it is provided. The operator will also perform a GET request on the user secret to get any credentials. 
+10. The operator will add the cluster certificate and any user related credentials to the secret it created earlier.
+11. The operator will add a `status.binding` to the `KafkaConnection` resource with the name of the secret. 
+12. The operator will then create `Watch` requests for the `Kafka` CR and, if used earlier, the `cluster-ca-cert` secret, the `KafkaUser` resource and the secret created by the `KafkaUser` operator. 
+13. If the `KafkaConnection` CR or any of the other watched resources change the operator will make the matching updates to the secret it created.
```

Comment:
Why is this approach better than completely separate operator in a completely separate repo? Given how long the current build pipeline runs we should try to look more into how to split things rather than add more.

I would also suspect that this has:
* Different release cycle (most Strimzi releases don't really change anything with regards to the `KafkaConnection`. Only new CRD versions might change it)
* Different installation mode (you cn have this for example installed as cluster wide while hacing the cluster operator installed as namespaced)

Reply:
I went for this approach initially as it matches the approach we have for other operator such as the KafkaConnect operator. Integrating it into the existing operator means we don't have to maintain a separate build and packaging system. However, I do think it's a good point that this will likely be on a different release cycle.

The point I hadn't considered was the different installation mode. Allowing the KafkaConnection operator to be cluster wide and the cluster operator to be namespaced I think is a big enough advantage that perhaps it outweighs the additional overhead.

At the moment the logic to handle reconcile loops and watches is all contained in the operator-common directory. We could look at creating a new project that still pulls in that code somehow. Alternatively we could look at operator-sdk. Operator-sdk doesn't currently support Java but this operator only has to watch Kubernetes resources, it doesn't have to talk to Kafka so there isn't really a reason why it has to be in Java.

Reply:
If we want to develop a new operator in Java then this is the way to go https://javaoperatorsdk.io/

Reply:
I hadn't realised there was a Java operator-sdk, I had heard rumours but it wasn't linked from the operator-sdk site so I didn't spot it

Reply:
> If we want to develop a new operator in Java then this is the way to go https://javaoperatorsdk.io/

Keep in mind that we need to integrate it with our existing resources. So you need it to be able to use our `api` extensively. Not sure how will that goes together or not.